File: python/second/subarray_sum_equals_k.py
from collections import defaultdict
class Solution(object):
    # O(n) time | O(n) space
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        curSum = 0
        res = 0
        freqMap = defaultdict(int)
        freqMap[0] = 1

        for num in nums:
            curSum += num
            diff = curSum - k
            res += freqMap[diff]
            freqMap[curSum] += 1
        return res


nums = [1,2,3]
k = 3

solution = Solution()
print(solution.subarraySum(nums, k))

File: src/main/java/november_2025/SubarraySumEqualsK.java
package november_2025;

import java.util.HashMap;
import java.util.Map;

public class SubarraySumEqualsK {

    public static void main(String[] args) {
        int[] nums = {1,2,3};
        int k = 3;

        SubarraySumEqualsK subarraySumEqualsK = new SubarraySumEqualsK();
        int result = subarraySumEqualsK.subarraySum(nums, k);
        System.out.println(result);

    }

    // O(n) time | O(n) space
    public int subarraySum(int[] nums, int k) {
        int res = 0;
        int curSum = 0;
        Map<Integer, Integer> prefixSum = new HashMap<>();
        prefixSum.put(0, 1);

        for (int num : nums) {
            curSum += num;
            int diff = curSum - k;
            res += prefixSum.getOrDefault(diff, 0);
            prefixSum.put(curSum, prefixSum.getOrDefault(curSum, 0) + 1);
        }
        return res;
    }

}
